modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C

```diff
@@ -0,0 +1,63 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ExposedSideAverageValue.h"
+#include "SelfShadowSideUserObject.h"
+
+registerMooseObject("MooseApp", ExposedSideAverageValue);
+
+InputParameters
+ExposedSideAverageValue::validParams()
+{
+  InputParameters params = SideAverageValue::validParams();
+  params.addClassDescription("Computes the average value of a variable on the "
+                             "exposed portion of a sideset. Note that this cannot be used on the "
+                             "centerline of an axisymmetric model.");
+  params.addRequiredParam<UserObjectName>(
+      "self_shadow_uo",
+      "SelfShadowSideUserObject that calculates the illumination state of a side set");
+  return params;
+}
+
+ExposedSideAverageValue::ExposedSideAverageValue(const InputParameters & parameters)
+  : SideAverageValue(parameters),
+    _self_shadow(getUserObject<SelfShadowSideUserObject>("self_shadow_uo"))
+{
+  if (_self_shadow.useDisplacedMesh() != getParam<bool>("use_displaced_mesh"))
+    paramError("use_displaced_mesh",
+               "The SelfShadowSideUserObject should operate on the same mesh (displaced or "
+               "undisplaced) as this PostProcessor.");
+}
+
+Real
+ExposedSideAverageValue::computeQpIntegral()
+{
+  const SelfShadowSideUserObject::SideIDType id(_current_elem->id(), _current_side);
+  const unsigned int illumination = _self_shadow.illumination(id);
+  // tests if the bit at position _qp is set
+  if (illumination & (1 << _qp))
```

Comment:
this `1 << _qp` thing does not seem right. where did it come from? would you have a minute to explain it to me?

Reply:
It is right. `<<` is the left shift operator. It generates the values 1 2 4 8 for _qp 0 1 2 3 (which are bit masks to test bit 0 1 2 3). The illumination state of all QPs of a side element is stored as a bitmask.

Reply:
I believe this is implemented correctly like @dschwen commented. However, I also think this is a confusing way to get that information from `SelfShadowSideUserObject`. I didn't really want to touch the implementation of `SelfShadowSideUserObject` in this PR. Maybe that could be improved in a follow-on PR?